pykoi/retrieval/vectordb/epsilla.py

```diff
@@ -0,0 +1,128 @@
+import os
+import numpy as np
+import types
+
+from typing import List, Union
+from langchain.embeddings.base import Embeddings
+from langchain.schema import BaseRetriever, Document
+from sklearn.decomposition import PCA
+from pyepsilla import vectordb
+
+from pykoi.retrieval.llm.constants import LlmName
+from pykoi.retrieval.vectordb.abs_vectordb import AbsVectorDb
+
+model_map = {
+    LlmName.OPENAI: 1536
+}
+
+class EpsillaRetriever(BaseRetriever):
+    def __init__(self, vector_db, embedding, limit):
+        self.vector_db = vector_db
+        self.embedding = embedding
+        self.limit = limit
+
+    def get_relevant_documents(self, query: str) -> List[Document]:
+        query_vector = self.embedding.embed_query(query)
+        status_code, response = self.vector_db.query(
+            table_name="RetrievalQA",
+            query_field="embeddings",
+            query_vector=query_vector,
+            limit=self.limit
+        )
+        if status_code != 200:
+            print(f"Error: {response['message']}.")
+            raise Exception("Error: {}.".format(response['message']))
+
+        return list(map(lambda item: Document(page_content = item["text"], metadata = dict()), response["result"]))
+
+    async def aget_relevant_documents(self, query: str) -> List[Document]:
+        # This is a placeholder for the asynchronous version of get_relevant_documents
+        # The actual implementation would depend on how the vectordb.Client and the embedding object handle
+        # asynchronous operations
+        pass
+
+class Epsilla(AbsVectorDb):
+    def __init__(self, model_name: Union[str, LlmName], embedding: Embeddings):
+        self._embedding = embedding
+        self._vector_db = vectordb.Client(
+            host="localhost",
+            port=8888,
```

Comment:
nit: can we add argument to allow user to pass in port number and use 8888 as default in case this port is taken.

Reply:
for now our web server use 8888, let me check a way to change it configurable in our db side first then modify this.